src/crimson/net/ProtocolV2.cc

```diff
@@ -481,16 +513,17 @@ seastar::future<> ProtocolV2::handle_auth_reply()
           logger().debug("{} auth reply more len={}",
                          conn, auth_more.auth_payload().length());
           ceph_assert(messenger.get_auth_client());
-          ceph::bufferlist reply;
-          int r = messenger.get_auth_client()->handle_auth_reply_more(
-               conn.shared_from_this(), auth_meta, auth_more.auth_payload(), &reply);
-          if (r < 0) {
+          try {
+            auto reply = messenger.get_auth_client()->handle_auth_reply_more(
+               conn.shared_from_this(), auth_meta, auth_more.auth_payload());
+            auto more_reply = AuthRequestMoreFrame::Encode(reply);
+            return write_frame(more_reply);
+          } catch (const std::system_error& e){
```

Comment:
i understand this change alters the behavior of this function. but shall we continue on even if `handle_auth_reply_more()` fails? if yes, does it make sense to send out the "reply" returned by a failed `handle_auth_reply_more()` call?

Reply:
We should not continue to `Encode()` nor `write_frame()` if `handle_auth_reply_more()` fails (originally return < 0, and now throw `error::negotiation_failure`), and the next logic should go to `fault()`. That's why `abort_in_fault()` is called here. Also see https://github.com/ceph/ceph/blob/master/src/msg/async/ProtocolV2.cc#L1785-L1789 


And I think `AuthClient::handle_auth_reply_more()` and `AuthClient::get_auth_request()` are better to use its own error/exception, maybe define them in `src/crimson/auth/`. It looks strange to me to throw `error::negotiation_failure` from non-msgr components. And the fact that `error::negotiation_failure` == "Tthe protocol should be aborted with fault()" is only implementation details of `Protocol` classes, should not become the responsibility of `AuthClient`.

Reply:
ahh, now i understand your reasoning here. i didn't realize that `abort_in_fault()` is a thin wrapper which simply throws an exception.

but i still don't think we should have something like
```c++
result_t result;
try {
  result = function_which_throws();
} catch (...) {
  error_handling();
}
continue_on_with(result);
```
which is counter intuitive, IMHO.

w.r.t. throwing `negotiation_failure` from non-msgr components, i don't fully agree with you. the components under `auth/` directory is the workhorse who offer the facilities to do authentication and authorization. and it's `mon::Client` which performs the authentication and authorization as a client or as a server. if there's any one who claims "we fail to be authenticated as osd.x", that'd be `mon::Client`.

msgr structures/parses the message frames for its consumers. so i think it should be `mon::Client` who throws the `negotiation_failure` exception. and msgr should be able to catch it, handle it and rethrow it if necessary.

> And the fact that `error::negotiation_failure` equals "The protocol should be aborted with fault()" is only implementation details of `Protocol` classes, should not become the responsibility of `AuthClient`.

agreed. that's why this exception is caught in msgr. and i think this exception should be part of the interface exposed by either `common/` or `auth/`.

Reply:
> but i still don't think we should have something like
```c++
result_t result;
try {
  result = function_which_throws();
} catch (...) {
  error_handling();
}
continue_on_with(result);
```
For `handle_auth_reply_more()`, we can just remove the try-catch-block and leave it to the fault handler https://github.com/ceph/ceph/blob/3b9a01653979fca925ca86b0417a369484ecb27e/src/crimson/net/ProtocolV2.cc#L818 which is the default way to handle failures, and add a comment here.

But for the below `get_auth_request()`, we might still need extra error_handling logic, because the closing and reset logic is special. See https://github.com/ceph/ceph/blob/master/src/msg/async/ProtocolV2.cc#L1714-L1720 

What do you think?

Reply:
great! i will remove the try-catch. and let the `fault()` in `execute_connecting()` to take care of the error handling.